runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/FieldAccessVisitor.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.beam.sdk.Pipeline.PipelineVisitor;
+import org.apache.beam.sdk.runners.TransformHierarchy.Node;
+import org.apache.beam.sdk.schemas.FieldAccessDescriptor;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.ParDo.MultiOutput;
+import org.apache.beam.sdk.util.Preconditions;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+
+/** Computes which Schema fields are (or conversely, are not) accessed in a pipeline. */
+class FieldAccessVisitor extends PipelineVisitor.Defaults {
+  private final Map<String, FieldAccessDescriptor> pCollectionFieldAccess = new HashMap<>();
+
+  /**
+   * Returns a map keyed by the {@link TupleTag} ID referencing a PCollection. Values are the set of
+   * fields accessed on that PCollection.
+   */
+  ImmutableMap<String, FieldAccessDescriptor> getPCollectionFieldAccess() {
+    return ImmutableMap.copyOf(pCollectionFieldAccess);
+  }
+
+  @Override
+  public void visitPrimitiveTransform(Node node) {
+    Map<String, FieldAccessDescriptor> currentFieldAccess = getFieldAccess(node);
+    for (Entry<String, FieldAccessDescriptor> entry : currentFieldAccess.entrySet()) {
+      FieldAccessDescriptor previousFieldAccess = pCollectionFieldAccess.get(entry.getKey());
+      FieldAccessDescriptor newFieldAccess =
+          previousFieldAccess == null
+              ? entry.getValue()
+              : FieldAccessDescriptor.union(
+                  ImmutableList.of(previousFieldAccess, entry.getValue()));
+      pCollectionFieldAccess.put(entry.getKey(), newFieldAccess);
+    }
+  }
+
+  private static Map<String, FieldAccessDescriptor> getFieldAccess(Node node) {
+    PTransform<?, ?> transform = node.getTransform();
+    HashMap<String, FieldAccessDescriptor> access = new HashMap<>();
+
+    if (transform instanceof MultiOutput) {
+      DoFn<?, ?> fn = ((MultiOutput<?, ?>) transform).getFn();
+      Pair<TupleTag<?>, PCollection<?>> mainInput = getMainInputTagId(node);
+      FieldAccessDescriptor fields =
+          ParDo.getDoFnSchemaInformation(fn, mainInput.getRight()).getFieldAccessDescriptor();
+      access.put(mainInput.getLeft().getId(), fields);
+    }
+
+    // For every input without field access info, we must assume all fields need to be accessed.
+    for (TupleTag<?> tag : node.getInputs().keySet()) {
+      if (!access.containsKey(tag.getId())) {
+        access.put(tag.getId(), FieldAccessDescriptor.withAllFields());
+      }
+    }
+
+    return ImmutableMap.copyOf(access);
+  }
+
+  private static Pair<TupleTag<?>, PCollection<?>> getMainInputTagId(Node node) {
```

Comment:
nit: This is also deceptively named, it appears to actually be `getMainInput`. (It gets both the PCollection and tagid.)

Reply:
Removed per your other comment.